fabfile.py

```diff
@@ -46,258 +68,281 @@
+
+        # Download database dump into VM
+        escaped_url = pipes.quote(config["url"])
+        run("wget " + escaped_url + " -O ~/dbdump")
+
+    # HACK: detect the Postgres user used in the dump. We run strings in case
+    # the dump is in binary format, then we look for the grant for arbitrary
+    # table, program_class. The result should be a line like:
+    #
+    #   GRANT ALL ON TABLE program_clas TO esp;
+    #
+    # ...which we can then parse to get the user. :D
+    contents = run("strings dbdump | grep 'GRANT ALL ON TABLE program_class TO'")
+    pg_owner = contents.split()[-1][:-1]
+
+    # Reset the database
+    emptydb(pg_owner, interactive=False)
+
+    # Load the database dump using the appropriate command for the format
+    sudo("chgrp postgres ~/dbdump")
+    if "PostgreSQL custom database dump" in run("file ~/dbdump"):
+        sudo("pg_restore --verbose --dbname=" + pipes.quote(env.dbname) +
+             " --exit-on-error --jobs=2 ~/dbdump",
+             user="postgres")
+    else:
+        sudo("psql --dbname=" + pipes.quote(env.dbname) +
+             " --set='ON_ERROR_STOP=on' -f ~/dbdump",
+             user="postgres")
 
-@task
-def recreate_encrypted_partition():
-    """ Blow away any previous encrypted partition and create a new one. """
+    # Run Django migrations, etc.
+    refresh()
 
-    create_encrypted_partition()
+    # Cleanup
+    run("rm -f ~/dbdump")
 
-@task
-def vagrant_dev_setup(dbuser=None, dbfile=None):
-    """ Set up the development environment.
-        If the dbuser and dbfile arguments are supplied, sets up
-        encrypted database storage and loads a DB dump. """
-
-    if dbuser is not None and dbfile is None:
-        raise Exception('You must provide a database dump file to load in the dbfile argument.')
-    if dbuser is None and dbfile is not None:
-        raise Exception('You must specify the PostgreSQL user that your database belongs to in the dbuser argument.')
-    using_db_dump = (dbuser is not None and dbfile is not None)
-
-    if using_db_dump:
-        load_db_dump(dbuser, dbfile)
-    else:
-        create_settings()
-        initialize_db()
-    setup_apache()
-    link_media()
+def gen_password(length):
+    return "".join([random.choice(string.letters + string.digits) for i in range(length)])
 
 @task
-def run_devserver():
-    """ Run Django dev server on port 8000. """
-    with esp_env():
-        sudo('python manage.py runserver 0.0.0.0:8000')
+def refresh():
+    """
+    Re-synchronize the remote environment with the codebase. For use when
+    switching branches or jumping around the history in git.
+
+      - upgrades and downgrades Python packages
+      - removes orphaned *.pyc files
+      - runs Django database migrations
+    """
+    ensure_environment()
+
+    with cd(env.rbase):
+        # Update the virtualenv with correct packages
+        run("pip install --upgrade -r " + env.rbase + "esp/requirements.txt")
```

Comment:
I like that this command does things relevant to both update_deps and manage.py update. Is there a reason why it does Python dependencies but not Ubuntu packages?